runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -359,21 +359,12 @@ static TR_OutlinedInstructions *generateArrayletReference(
 
    if (comp->target().is64Bit() && comp->useCompressedPointers())
       {
-      if (TR::Compiler->vm.heapBaseAddress() == 0)
-         isHB0 = true;
+      isHB0 = true;
```

Comment:
Can we fold uses of this as true?

Reply:
This is only `true` under compressedRefs on a 64-bit machine. We use this on line 463 below:
```
      if (loadNeedsDecompression)
         {
         if (comp->target().is64Bit() && comp->useCompressedPointers())

         if (isHB0)
            {
            if (shiftOffset > 0)
               {
               generateRegImmInstruction(SHL8RegImm1, node, loadOrStoreReg, shiftOffset, cg);
               }
            }
         else
            {
            TR_ASSERT(0, "!HB0 not supported yet");
            }
```

So I don't think we can fold this as true always.

Reply:
All the paths it is used under is guarded with an `useCompressedPointers()` check though, so in effect all the locations where we query this values it is always true. Do you agree? Also the a name "HB0" no longer makes sense, and future readers will be quite confused as to what that even means.

Reply:
The code is guarded by the `loadNeedsDecompression` boolean. This is defined as:
```
if (loadOrStoreOrArrayElementNode->getOpCodeValue() == TR::l2a || 
   ((loadOrStoreOrArrayElementNode->getOpCodeValue() == TR::aload || loadOrStoreOrArrayElementNode->getOpCodeValue() == TR::aRegLoad) && node->isSpineCheckWithArrayElementChild()) &&
       comp->target().is64Bit() && comp->useCompressedPointers())
      loadNeedsDecompression = true;
```

So if `loadOrStoreOrArrayElementNode->getOpCodeValue() == TR::l2a` is true, then `loadNeedsDecompression` will also be true, regardless of whether we are using compressed pointers, right?

Reply:
I don't think I explained my above comment sufficiently. What I meant is, we can end up in the following block:
```
      if (loadNeedsDecompression)
         {
         if (comp->target().is64Bit() && comp->useCompressedPointers())

         if (isHB0)
            {
            if (shiftOffset > 0)
               {
               generateRegImmInstruction(SHL8RegImm1, node, loadOrStoreReg, shiftOffset, cg);
               }
            }
         else
            {
            TR_ASSERT(0, "!HB0 not supported yet");
            }
```

Even if `useCompressedPointers()` is not true. So if we fold away `isHB0` as true, then we would indirectly assume that `loadNeedsDecompression` is also only ever true under compressed references. I agree though that the name `HB0` no longer makes sense. Perhaps we need to have a `bool is64BitCompressedRefs = comp->target().is64Bit() && comp->useCompressedPointers()` as this is used multiple times and can replace `isHB0` as well. Let me know what you think

Reply:
I'm not following why `loadNeedsDecompression` has anything to do with folding `isHB0`. The logic behind folding `isHB0` is simple:

- `isHB0` is set to true if and only if `TR::Compiler->vm.heapBaseAddress() == 0` **AND** `comp->target().is64Bit() && comp->useCompressedPointers()`
	- Heap base is always 0 so the former statement is always true
- Any uses of `isHB0` which are guarded by `comp->target().is64Bit() && comp->useCompressedPointers()` is therefore always true as well
	- There is only one other use of `isHB0` on line 472, and it is indeed guarded by `comp->target().is64Bit() && comp->useCompressedPointers(), therefore it can be folded as true

The logic follows that `isHB0` variable is true on all paths in which it is used, therefore we can just get rid of it. Does that make sense? Notice `loadNeedsDecompression` never came into the picture of such logic as whether this variable is true or not is irrelevant.

Reply:
What is slightly confusing is that the original code does not use braces. I'll paste the original code here:

```
      if (loadNeedsDecompression)
         {
         if (comp->target().is64Bit() && comp->useCompressedPointers())

         if (isHB0)
            {
            if (shiftOffset > 0)
               {
               generateRegImmInstruction(SHL8RegImm1, node, loadOrStoreReg, shiftOffset, cg);
               }
            }
         else
            {
            TR_ASSERT(0, "!HB0 not supported yet");
            }
```

And how I would have wrote it:
```
      if (loadNeedsDecompression)
         {
         if (comp->target().is64Bit() && comp->useCompressedPointers())
            {
            if (isHB0)
               {
               if (shiftOffset > 0)
                  {
                  generateRegImmInstruction(SHL8RegImm1, node, loadOrStoreReg, shiftOffset, cg);
                  }
               }
            else
               {
               TR_ASSERT(0, "!HB0 not supported yet");
               }
            }
```

Does the braces and indentation help to understand my above logic?

Reply:
Ah I see the issue. This line `if (comp->target().is64Bit() && comp->useCompressedPointers())` was removed from my branch because it saw it as redundant with `isHB0`. 

I think it's easiest if I just restore the original `if (comp->target().is64Bit() && comp->useCompressedPointers())` and remove `isHB0`.

Reply:
2b7638c96a0bf1c5e17d228ba8ab3eaada141f3d